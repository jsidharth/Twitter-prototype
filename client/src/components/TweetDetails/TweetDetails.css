.flexTweetDetails {
  display: flex;
  flex-direction: row;
}

.tweetDetailsContainer {
  margin-left: 50px;
}

.cardContainer {
  width: 600px;
}

.cardWidth {
  width: 600px;
}

.paperHeight {
  padding-top: 10px;
  height: 50px;
  font-weight: 800;
  font-size: 19px;
  font-family: system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Ubuntu,
    'Helvetica Neue', sans-serif;
  padding-left: 15px;
  border-style: solid;
  border-width: 0.5px;
  border-color: #d0d4d6;
  display: flex;
  flex-direction: row;
}

.backIconTweetDetails {
    margin-right: 20px;
    color: #00acee;
}

.profileImageTweetDetails {
  margin: 0 auto;
  max-width: 80px;
  max-height: 80px;
  border-radius: 50%;
}

.tweetBody {
  margin-left: 10px;
}