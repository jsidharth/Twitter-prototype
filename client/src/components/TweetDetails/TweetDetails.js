/* eslint-disable react/destructuring-assignment */
/* eslint-disable react/prop-types */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import './TweetDetails.css';
import { FiMessageCircle } from 'react-icons/fi';
import { MdKeyboardBackspace, MdBookmarkBorder } from 'react-icons/md';
import { FaRegHeart } from 'react-icons/fa';
import { AiOutlineRetweet } from 'react-icons/ai';
import { Link } from 'react-router-dom';
import { tweetActions } from '../../js/actions/index';
import Sidebar from '../Sidebar/Sidebar';
import CommentModal from '../CommentModal/CommentModal';
import TweetCard from '../TweetCard/TweetCard';

class TweetDetails extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showCommentModal: false,
      tweet: {}
    };
    this.likeTweet = this.likeTweet.bind(this);
    this.unlikeTweet = this.unlikeTweet.bind(this);
    this.bookmarkTweet = this.bookmarkTweet.bind(this);
    this.showCommentModal = this.showCommentModal.bind(this);
  }

  componentDidMount() {
    const data = {
      tweetId: this.props.match.params.tweetID,
    };

    const { getTweetDetails } = this.props;
    getTweetDetails(data);
  }

  componentDidUpdate(prevProps) {
    if (this.props.match.params.tweetID !== prevProps.match.params.tweetID) {
      const { getTweetDetails } = this.props;
      const data = {
        tweetId: this.props.match.params.tweetID,
      };
      getTweetDetails(data);
    }
  }

  // Render the tweet detail after a comment has been posted
  componentWillReceiveProps(nextProps) {
    const { tweet } = nextProps;
    this.setState({
      tweet
    });
  }

  likeTweet = e => {
    let data = { tweetId: e.target.id, userId: this.props.userId };
    this.props.likeTweet(data).then(() => {
      this.props.getTweetDetails(data);
    });
  };

  unlikeTweet = e => {
    let data = { tweetId: e.target.id, userId: this.props.userId };
    this.props.unlikeTweet(data).then(() => {
      this.props.getTweetDetails(data);
    });
  };

  bookmarkTweet = e => {
    let data = { tweetId: e.target.id, userId: this.props.userId };
    this.props.bookmarkTweet(data).then(() => {
      this.props.getTweetDetails(data);
    });
  };

  showCommentModal = () => {
    this.setState({
      showCommentModal: !this.state.showCommentModal,
    });
  };

  render() {
    const { tweet } = this.state;
    const tweetUserId = tweet.userId;
    var myDate;
    var timeValue;
    if (tweet.created_at) {
      myDate = new Date(tweet.created_at);
      myDate = myDate.toString().split(' ');
      timeValue = myDate[4].split(':');
      var time = myDate[4].split(':');
      var date = myDate[1] + ' ' + myDate[2] + ',' + myDate[3];
      time = timeValue[0] + ':' + timeValue[1];
    }

    let likeButton, unlikeButton, renderLikeButton;
    if (tweet.body) {
      likeButton = <FaRegHeart size={20} id={tweet._id} onClick={this.likeTweet} />;
      unlikeButton = (
        <svg
          xmlns="http://www.w3.org/2000/svg"
          width="20"
          height="20"
          viewBox="0 0 24 24"
          fill="red"
        >
          <path
            id={tweet._id}
            onClick={this.unlikeTweet}
            d="M12 4.435c-1.989-5.399-12-4.597-12 3.568 0 4.068 3.06 9.481 12 14.997 8.94-5.516 12-10.929 12-14.997 0-8.118-10-8.999-12-3.568z"
          />
        </svg>
      );
      renderLikeButton = tweet.likes.includes(this.props.userId) ? unlikeButton : likeButton;
    }
    const imgSrc = tweet.profilePic ? tweet.profilePic : '/images/default_profile_bigger.png';

    return (
      <div className="flexTweetDetails">
        <Sidebar />
        <div>
          {this.state.showCommentModal ? (
            <CommentModal
              showCommentModal={this.showCommentModal}
              showCommentModalState={this.state.showCommentModal}
              tweetProfilePic={tweet.profilePic}
              tweetUserName={tweet.name}
              tweetUserHandle={tweet.handle}
              tweetDate={date}
              tweetBody={tweet.body}
              tweetUserId={tweetUserId}
              tweetId={tweet._id}
              currentUser={this.props.user}
            />
          ) : null}
          <div className="cardContainer">
            <div className="cardWidth">
              <div className="paperHeight">
                <Link to="/home">
                  <div className="backIconTweetDetails">
                    <MdKeyboardBackspace size={30} />
                  </div>
                </Link>
                Tweet
              </div>
            </div>
          </div>
          <div className="cardContent">
            <div className="flexImageTweet">
              <div>
                {/* Include user profile image if available */}
                <img src={imgSrc} className="profileImageTweetDetails" alt="user" />
              </div>
              <div>
                <div className="flexNameHandle">
                  <p className="tweetUserName">{tweet.name}</p>
                  <p className="tweetUserHandle">@{tweet.handle}</p>
                </div>
                <p className="tweetBody">{tweet.body}</p>
                {tweet.image ? <img src={tweet.image} alt="Tweet" className="tweetImage" /> : null}
              </div>
            </div>
            <div className="flexDateTime">
              <p className="tweetTimeDetails">{time}</p>
              <p className="tweetDateDetails">{date}</p>
            </div>
            <div className="flexCardBtns">
              <div className="flexBtnCnt">
                <FiMessageCircle size={20} onClick={this.showCommentModal} />
                <div>{tweet.comments_count > 0 ? tweet.comments_count : null}</div>
              </div>
              <div className="flexBtnCnt">
                <AiOutlineRetweet size={20} />
                <div>{tweet.retweet_count > 0 ? tweet.retweet_count : null}</div>
              </div>
              <div className="flexBtnCnt">
                <div>{renderLikeButton}</div>
                <div>{tweet.likes_count > 0 ? tweet.likes_count : null}</div>
              </div>
              <div className="flexBtnCnt">
                <MdBookmarkBorder size={20} id={tweet._id} onClick={this.bookmarkTweet} />
              </div>
            </div>
          </div>
          {tweet.comments && tweet.comments.length ? (
            <TweetCard
              tweets={tweet.comments}
              likeTweet={this.likeTweet}
              unlikeTweet={this.unlikeTweet}
            />
          ) : null}
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  tweet: state.tweet.currentTweet,
  userId: state.user.currentUser._id,
  user: state.user.currentUser,
});

const mapDispatchToProps = dispatch => ({
  getTweetDetails: data => dispatch(tweetActions.getTweetDetails(data)),
  likeTweet: data => dispatch(tweetActions.likeTweet(data)),
  unlikeTweet: data => dispatch(tweetActions.unlikeTweet(data)),
  bookmarkTweet: data => dispatch(tweetActions.bookmarkTweet(data)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TweetDetails);
