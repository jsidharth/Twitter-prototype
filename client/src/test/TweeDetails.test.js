// Link.react.test.js
import React from 'react';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import TweetDetails from '../components/TweetDetails/TweetDetails';

configure({ adapter: new Adapter() });

it('render correctly the TweetDetails component', () => {
  const tweetDetailsComponent = renderer.create(<TweetDetails />).toJSON();
  expect(tweetDetailsComponent).toMatchSnapshot();
});
