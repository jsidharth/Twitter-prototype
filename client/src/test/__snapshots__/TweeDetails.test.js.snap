// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly the TweetDetails component 1`] = `
<div
  className="flexTweetDetails"
>
  <div
    className="StickySidebar"
  >
    <div>
      <nav
        className="sidebar"
      >
        <div
          className="list-group list-group-flush"
          onKeyDown={[Function]}
        >
          <div>
            <div
              className="list-group-item"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
            >
              <span
                className="navbar-brand"
              >
                <img
                  alt="Main logo link to home"
                  className="d-inline-block align-top"
                  height="35px"
                  src="https://upload.wikimedia.org/wikipedia/fr/thumb/c/c8/Twitter_Bird.svg/944px-Twitter_Bird.svg.png"
                  width="45px"
                />
              </span>
            </div>
          </div>
          <div>
            <div
              className="list-group-item"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
            >
              <div
                className="flexSidebars"
              >
                <svg
                  fill="currentColor"
                  height={30}
                  size={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 16 16"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M16 9l-3-3V2h-2v2L8 1 0 9h2l1 5c0 .55.45 1 1 1h8c.55 0 1-.45 1-1l1-5h2zm-4 5H9v-4H7v4H4L2.81 7.69 8 2.5l5.19 5.19L12 14z"
                    fillRule="evenodd"
                  />
                </svg>
                <p
                  className="sidebarTopics"
                >
                  Home
                </p>
              </div>
            </div>
          </div>
          <div>
            <div
              className="list-group-item"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
            >
              <div
                className="flexSidebars"
              >
                <svg
                  fill="currentColor"
                  height={30}
                  size={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 14 16"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 4v8c0 .55.45 1 1 1h12c.55 0 1-.45 1-1V4c0-.55-.45-1-1-1H1c-.55 0-1 .45-1 1zm13 0L7 9 1 4h12zM1 5.5l4 3-4 3v-6zM2 12l3.5-3L7 10.5 8.5 9l3.5 3H2zm11-.5l-4-3 4-3v6z"
                    fillRule="evenodd"
                  />
                </svg>
                <p
                  className="sidebarTopics"
                >
                  Messages
                </p>
              </div>
            </div>
          </div>
          <div>
            <div
              className="list-group-item"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
            >
              <div
                className="flexSidebars"
              >
                <svg
                  fill="none"
                  height={30}
                  size={30}
                  stroke="currentColor"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 24 24"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M19 21l-7-5-7 5V5a2 2 0 0 1 2-2h10a2 2 0 0 1 2 2z"
                  />
                </svg>
                <p
                  className="sidebarTopics"
                >
                  Bookmarks
                </p>
              </div>
            </div>
          </div>
          <div>
            <div
              className="list-group-item"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
            >
              <div
                className="flexSidebars"
              >
                <svg
                  fill="currentColor"
                  height={30}
                  size={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 14 16"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M3 10h4V9H3v1zm0-2h6V7H3v1zm0-2h8V5H3v1zm10 6H1V3h12v9zM1 2c-.55 0-1 .45-1 1v9c0 .55.45 1 1 1h12c.55 0 1-.45 1-1V3c0-.55-.45-1-1-1H1z"
                    fillRule="evenodd"
                  />
                </svg>
                <p
                  className="sidebarTopics"
                >
                  Lists
                </p>
              </div>
            </div>
          </div>
          <div>
            <div
              className="list-group-item"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
            >
              <div
                className="flexSidebars"
              >
                <svg
                  fill="currentColor"
                  height={30}
                  size={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 496 512"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M248 104c-53 0-96 43-96 96s43 96 96 96 96-43 96-96-43-96-96-96zm0 144c-26.5 0-48-21.5-48-48s21.5-48 48-48 48 21.5 48 48-21.5 48-48 48zm0-240C111 8 0 119 0 256s111 248 248 248 248-111 248-248S385 8 248 8zm0 448c-49.7 0-95.1-18.3-130.1-48.4 14.9-23 40.4-38.6 69.6-39.5 20.8 6.4 40.6 9.6 60.5 9.6s39.7-3.1 60.5-9.6c29.2 1 54.7 16.5 69.6 39.5-35 30.1-80.4 48.4-130.1 48.4zm162.7-84.1c-24.4-31.4-62.1-51.9-105.1-51.9-10.2 0-26 9.6-57.6 9.6-31.5 0-47.4-9.6-57.6-9.6-42.9 0-80.6 20.5-105.1 51.9C61.9 339.2 48 299.2 48 256c0-110.3 89.7-200 200-200s200 89.7 200 200c0 43.2-13.9 83.2-37.3 115.9z"
                  />
                </svg>
                <p
                  className="sidebarTopics"
                >
                  Profile
                </p>
              </div>
            </div>
          </div>
          <div>
            <div
              className="list-group-item"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
            >
              <div
                className="flexSidebars"
              >
                <svg
                  fill="currentColor"
                  height={30}
                  size={30}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 16 16"
                  width={30}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M16 14v1H0V0h1v14h15zM5 13H3V8h2v5zm4 0H7V3h2v10zm4 0h-2V6h2v7z"
                    fillRule="evenodd"
                  />
                </svg>
                <p
                  className="sidebarTopics"
                >
                  Analytics
                </p>
              </div>
            </div>
          </div>
          <div>
            <div
              className="list-group-item"
              data-rb-event-key={null}
              disabled={false}
              onClick={[Function]}
            >
              <div
                className="Signout"
                id="signout"
                onClick={[Function]}
              >
                <div
                  className="flexSidebars"
                >
                  <svg
                    fill="none"
                    height={30}
                    size={30}
                    stroke="currentColor"
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth="2"
                    style={
                      Object {
                        "color": undefined,
                      }
                    }
                    viewBox="0 0 24 24"
                    width={30}
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
                    />
                    <polyline
                      points="16 17 21 12 16 7"
                    />
                    <line
                      x1="21"
                      x2="9"
                      y1="12"
                      y2="12"
                    />
                  </svg>
                  <p
                    className="sidebarTopics"
                  >
                    Logout
                  </p>
                </div>
              </div>
            </div>
          </div>
          <button
            className="tweetBtn"
            onClick={[Function]}
            type="button"
          >
            Tweet
          </button>
        </div>
      </nav>
    </div>
  </div>
  <div>
    <div
      className="cardContainer"
    >
      <div
        className="cardWidth"
      >
        <div
          className="paperHeight"
        >
          <div
            className="backIconTweetDetails"
          >
            <svg
              fill="currentColor"
              height={30}
              size={30}
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width={30}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 11H6.83l3.58-3.59L9 6l-6 6 6 6 1.41-1.41L6.83 13H21z"
              />
            </svg>
          </div>
          Tweet
        </div>
      </div>
    </div>
    <div
      className="cardContent"
    >
      <div
        className="flexImageTweet"
      >
        <div>
          <img
            alt="user"
            className="profileImageTweetDetails"
            src="/images/default_profile_bigger.png"
          />
        </div>
        <div>
          <div
            className="flexNameHandle"
          >
            <p
              className="tweetUserName"
            />
            <p
              className="tweetUserHandle"
            >
              @
            </p>
          </div>
          <p
            className="tweetBody"
          />
        </div>
      </div>
      <div
        className="flexDateTime"
      >
        <p
          className="tweetTimeDetails"
        />
        <p
          className="tweetDateDetails"
        />
      </div>
      <div
        className="flexCardBtns"
      >
        <div
          className="flexBtnCnt"
        >
          <svg
            fill="none"
            height={20}
            onClick={[Function]}
            size={20}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width={20}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M21 11.5a8.38 8.38 0 0 1-.9 3.8 8.5 8.5 0 0 1-7.6 4.7 8.38 8.38 0 0 1-3.8-.9L3 21l1.9-5.7a8.38 8.38 0 0 1-.9-3.8 8.5 8.5 0 0 1 4.7-7.6 8.38 8.38 0 0 1 3.8-.9h.5a8.48 8.48 0 0 1 8 8v.5z"
            />
          </svg>
          <div />
        </div>
        <div
          className="flexBtnCnt"
        >
          <svg
            fill="currentColor"
            height={20}
            onClick={[Function]}
            size={20}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 1024 1024"
            width={20}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M136 552h63.6c4.4 0 8-3.6 8-8V288.7h528.6v72.6c0 1.9.6 3.7 1.8 5.2a8.3 8.3 0 0 0 11.7 1.4L893 255.4c4.3-5 3.6-10.3 0-13.2L749.7 129.8a8.22 8.22 0 0 0-5.2-1.8c-4.6 0-8.4 3.8-8.4 8.4V209H199.7c-39.5 0-71.7 32.2-71.7 71.8V544c0 4.4 3.6 8 8 8zm752-80h-63.6c-4.4 0-8 3.6-8 8v255.3H287.8v-72.6c0-1.9-.6-3.7-1.8-5.2a8.3 8.3 0 0 0-11.7-1.4L131 768.6c-4.3 5-3.6 10.3 0 13.2l143.3 112.4c1.5 1.2 3.3 1.8 5.2 1.8 4.6 0 8.4-3.8 8.4-8.4V815h536.6c39.5 0 71.7-32.2 71.7-71.8V480c-.2-4.4-3.8-8-8.2-8z"
            />
          </svg>
          <div />
        </div>
        <div
          className="flexBtnCnt"
        >
          <div />
          <div />
        </div>
        <div
          className="flexBtnCnt"
        >
          <svg
            fill="currentColor"
            height={20}
            onClick={[Function]}
            size={20}
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width={20}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 3H7c-1.1 0-1.99.9-1.99 2L5 21l7-3 7 3V5c0-1.1-.9-2-2-2zm0 15l-5-2.18L7 18V5h10v13z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
